### ISSUE TITLE
Unnecessary re-render when using the hooks API

### ISSUE THREAD

**Issue by PepijnSenders** [facebook/react#16177]
**Do you want to request a *feature* or report a *bug*?**

Asking for advice.

**What is the current behavior?**

When you have a component that conditionally depends on a hook, it's quite hard to extract this logic to a place and not have the app do unnecessary re-renders.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than React. Paste the link to your JSFiddle (https://jsfiddle.net/Luktwrdm/) or CodeSandbox (https://codesandbox.io/s/new) example below:**

Implementation with hooks: https://codesandbox.io/s/nifty-williams-jg3q2
Without hooks: https://codesandbox.io/s/pedantic-ganguly-xchuz

**What is the expected behavior?**

We could, of course, send a value down to the hook `useLongOperation({ dontRun: true })`, but this seems quite ugly to me. I expect some kind of API to do this more cleanly.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**

In previous versions of React you would just do the long operation on a didMount lifecycle when the context value equals true. The component would then have the images inside and would just re-render the image with the changed opacity.

**Comment by dangcuuson** [comment]
It'd be nice if you could provide a version using class component for comparison.

It's clear that all the re-render happens because `App` states changed when you press the button.

If you don't want to pass `dontRun` to your custom hook, you can directly `useContext` inside `useLongOperation`, which I believe is more similar to the class component implementation that you described (unless you'd want to re-use the custom hook in another place)

**Comment by PepijnSenders** [comment]
Passing `dontRun` and using the context directly in the hook are basically the same thing. The thing is that in this example the extracted logic in `useLongOperation` is just used in one place, but in a real-world example, it would be used in a dozen places. What I want to accomplish is an elegant way of being able to use the same component, in both cases of the hook being used and not.

**Comment by PepijnSenders** [comment]
I'll write up an example in "old" React.

**Comment by PepijnSenders** [comment]
This is the "old" react example: https://codesandbox.io/s/pedantic-ganguly-xchuz. When you run the app you can see that the images don't reload. This is because it uses the same component reference for those elements and manages to reuse its state in the DOM tree. This is sadly not possible if you use two different components.

**Comment by dangcuuson** [comment]
From what I see, the example with hooks uses different component structure than the class example.

In the class example, it's always `App`->`Kittens`->`img`

Whereas in the hooks example, it's `App`->`Kittens`->`KittenWithLongOp or KittenImages`->`img`

I think the conditional rendering in the hooks example is the cause for the images to be reloaded, because when you press the button, the `KittenImages` component is replaced by `KittenWithLongOp` in the component hierarchy, which makes React unmount all of its children (the `img`s in this case), and mount another `img`s.

Basically in class example, the `img` components is re-rendered, but in hooks example, the `img` components are unmounted & mount again.

I don't know the implementation detail of the `img` component so it could be another reason, but here's a hook implementation that is closer to the class component version, and I doesn't have the reload issue:

[https://codesandbox.io/s/focused-pascal-8mvlr]https://codesandbox.io/s/focused-pascal-8mvlr)

**Comment by PepijnSenders** [comment]
That is correct, and I also am aware of this. I'm just looking for an elegant way to not pass this value to the longOperation hook. Or be able to reuse the component structure for the images.
